tests/integration/relations/opensearch_provider/test_opensearch_provider.py

```diff
@@ -37,32 +52,153 @@ async def test_database_relation_with_charm_libraries(
+            opensearch_charm,
+            application_name=OPENSEARCH_APP_NAME,
+            num_units=NUM_UNITS,
+            series=SERIES,
         ),
         ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="edge", config=tls_config),
     )
+    await ops_test.model.relate(OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME)
+    wait_for_relation_joined_between(ops_test, OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME)
+
+    global client_relation
+    client_relation = await ops_test.model.add_relation(
+        f"{OPENSEARCH_APP_NAME}:{ClientRelationName}", f"{CLIENT_APP_NAME}:first-database"
+    )
+    wait_for_relation_joined_between(ops_test, OPENSEARCH_APP_NAME, CLIENT_APP_NAME)
+
+    async with ops_test.fast_forward():
+        # This test shouldn't take so long
+        await ops_test.model.wait_for_idle(timeout=1200, status="active")
+
+
+@pytest.mark.client_relation
+async def test_database_usage(ops_test: OpsTest):
+    """Check we can update and delete things."""
+    run_create_index = await run_simple_put(
+        ops_test,
+        unit_name=ops_test.model.applications[CLIENT_APP_NAME].units[0].name,
+        relation_id=client_relation.id,
+    )
+    # TODO have better validation here.
+    logging.error(json.dumps(run_create_index))
+
+    read_index_endpoint = "/albums/_search?q=Jazz"
+    run_read_index = await run_get_from_index(
+        ops_test,
+        unit_name=ops_test.model.applications[CLIENT_APP_NAME].units[0].name,
+        endpoint=read_index_endpoint,
+        relation_id=client_relation.id,
+    )
+    results = json.loads(run_read_index["results"])
+    logging.error(results)
+    assert results.get("timed_out") is False
+    assert results.get("hits", {}).get("total", {}).get("value") == 1
+
+
+@pytest.mark.client_relation
+async def test_database_bulk_usage(ops_test: OpsTest):
+    """Check we can update and delete things using bulk api."""
+    run_bulk_create_index = await run_bulk_put(
+        ops_test,
+        unit_name=ops_test.model.applications[CLIENT_APP_NAME].units[0].name,
+        relation_id=client_relation.id,
+    )
+    # change assertion to "data written" or something
+    logging.info(json.dumps(run_bulk_create_index["results"]))
+
+    read_index_endpoint = "/albums/_search?q=Jazz"
+    run_bulk_read_index = await run_get_from_index(
+        ops_test,
+        unit_name=ops_test.model.applications[CLIENT_APP_NAME].units[0].name,
+        endpoint=read_index_endpoint,
+        relation_id=client_relation.id,
+    )
+    # TODO assert we're getting the correct value
+    results = json.loads(run_bulk_read_index["results"])
+    logging.error(results)
+    assert results.get("timed_out") is False
+    assert results.get("hits", {}).get("total", {}).get("value") == 3
+
+
+@pytest.mark.client_relation
+async def test_database_version(ops_test: OpsTest):
+    """Check version is accurate."""
+    run_version_query = await run_get_from_index(
```

Comment:
nit: I'm not sure the name `get_from_index` is explanatory in the case of non-data related operations, such as retrieving the health or the version etc.. what do you think?

Reply:
I've updated it to just be called `get_request` - how's that?

Reply:
I like it !